zephyr/include/sof/lib/regions_mm.h

```diff
@@ -10,31 +10,239 @@
+					  int memory_region_attribute, bool span);
+void *vmh_alloc(struct virtual_memory_heap *heap, uint32_t alloc_size);
+int vmh_free_heap(struct virtual_memory_heap *heap);
+int vmh_free(struct virtual_memory_heap *heap, void *ptr);
+struct virtual_memory_heap *vmh_reconfigure_heap(struct virtual_memory_heap *heap,
+						 struct heap_config *cfg, bool span);
+void vmh_get_default_heap_config(const struct sys_mm_drv_region *region, struct heap_config *cfg);
+struct virtual_memory_heap *vmh_get_heap_by_attribute(uint32_t attr, uint32_t core_id);
+/**
+ * @brief Checks if ptr is in range of given memory range
+ *
+ * @param ptr checked ptr
+ * @param range_start start of checked memory range
+ * @param range_size size of checked memory range.
+ *
+ * @retval false if no overlap detected.
+ * @retval true if any overlap detected.
+ */
+static inline bool vmh_is_ptr_in_memory_range(void *ptr, void *range_start, size_t range_size)
+{
+	return (uint8_t *)ptr >= (uint8_t *)range_start &&
+	       (uint8_t *)ptr < (uint8_t *)range_start + range_size;
+}
+
+/**
+ * @brief Checks if given memory range overlaps second range.
+ *
+ * @param checked_memory_ptr start of first checked memory range
+ * @param checked_memory_size size of first checked memory range.
+ * @param range_start start of memory range
+ * @param range_size size of range.
+ *
+ * @retval false if no overlap detected.
+ * @retval true if any overlap detected.
+ */
+static inline bool vmh_do_ranges_overlap(void *checked_memory_ptr, size_t checked_memory_size,
+					 void *range_start, size_t range_size)
+{
+	uint8_t *checked_memory_end = (uint8_t *)checked_memory_ptr + checked_memory_size;
+	uint8_t *range_end = (uint8_t *)range_start + range_size;
+
+	return ((uint8_t *)checked_memory_ptr >= (uint8_t *)range_start &&
+		(uint8_t *)checked_memory_ptr < range_end) ||
+		(checked_memory_end > (uint8_t *)range_start &&
+		checked_memory_end <= range_end);
+}
+
+/**
+ * @brief Checks if given memory range overlaps second range
+ * in full.
+ *
+ * @param checked_memory_ptr start of first checked memory range
+ * @param checked_memory_size size of first checked memory range.
+ * @param range_start start of range
+ * @param range_size size of memory range.
+ *
+ * @retval false if no overlap detected or checked memory is not
+ * fully encapsulated in range.
+ * @retval true if checked memory is fully in given range.
+ */
+static inline bool vmh_is_range_fully_in_range(void *checked_memory_ptr,
+					       size_t checked_memory_size, void *range_start,
+					       size_t range_size)
+{
+	uint8_t *checked_memory_end = (uint8_t *)checked_memory_ptr + checked_memory_size;
+	uint8_t *range_end = (uint8_t *)range_start + range_size;
+
+	return (uint8_t *)checked_memory_ptr >= (uint8_t *)range_start &&
+		(uint8_t *)checked_memory_ptr < range_end &&
+		checked_memory_end > (uint8_t *)range_start &&
+		checked_memory_end <= range_end;
```

Comment:
you anyway implicitly assume that ranges are valid, i.e. `start <= end` or even `start < end` so it would be enough here to do
```
return (uint8_t *)checked_memory_ptr >= (uint8_t *)range_start && checked_memory_end <= range_end;
```